deltadb/src/catalog_server.c

```diff
@@ -475,23 +477,71 @@ static void handle_query( struct link *ql, time_t st )
 		return;
 	}
 
-	if(sscanf(url, "http://%[^/]%s", hostport, path) == 2) {
+	if(sscanf(url, "http://%[^/]%s", hostport, full_path) == 2) {
 		// continue on
 	} else {
-		strcpy(path, url);
+		strcpy(full_path, url);
 	}
 
-	/* load the hash table entries into one big array */
+	if(3==sscanf(full_path, "/updates/%ld/%ld/%[^/]",&time_start,&time_stop,strexpr)) {
+		// check for updates request first, before processing deltadb state
+		struct buffer buf;
+		buffer_init(&buf);
+		if(b64_decode(strexpr,&buf)==0) {
+			struct jx *expr = jx_parse_string(buffer_tostring(&buf));
+			if(expr) {
+				if(link_using_ssl(ql)) {
+					send_http_response(ql,501,"Server Error","text/plain",st);
+					link_printf(ql,st,"Sorry, unable to serve queries over HTTPS.");
+				} else {
+					send_http_response(ql,200,"OK","text/plain",st);
 
+					struct deltadb_query *query = deltadb_query_create();
+					deltadb_query_set_filter(query,expr);
+					// Note this leaks a stdio stream, but it will be shortly recovered on process exit.
+					deltadb_query_set_output(query,fdopen(link_fd(ql),"w"));
+					deltadb_query_set_display(query,DELTADB_DISPLAY_STREAM);
+					deltadb_query_execute_dir(query,history_dir,time_start,time_stop);
+					deltadb_query_delete(query);
+					jx_delete(expr);
+				}
+			} else {
+				send_http_response(ql,400,"Bad Request","text/plain",st);
+				link_printf(ql,st,"Invalid query text.\n");
+			}
+		} else {
+			send_http_response(ql,400,"Bad Request","text/plain",st);
+			link_printf(ql,st,"Invalid base-64 encoding.\n");
+		}
+		buffer_free(&buf);
+		return;
+	}
+
+	// check for historical timestamp prefix
+	int matches = sscanf(full_path, "/history/%ld%s", &timestamp, path);
+	if (matches == 2) {
+		// continue on
+	} else if (matches == 1) {
+		strncpy(path, "/", sizeof("/"));
+	} else {
+		strcpy(path, full_path);
+	}
+
+	// reset catalog to timestamp
+	if (timestamp > 0) {
+		// deltadb_delete(table);
```

Comment:
NOTE:  unsure sure if not deleting `table` before overwriting it causes a leak, since the variable is inherited during process fork.  Either way, it might be better to avoid deleting since it impacts performance, and the leak will be recovered shortly after on process exit.